klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+# Copyright 2022 GlobalFoundries PDK Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Run GlobalFoundries 180nm MCU DRC Unit Regression.
+
+Usage:
+    run_regression.py (--help| -h)
+    run_regression.py [--mp=<num>] [--run_name=<run_name>] [--table_name=<table_name>]
+
+Options:
+    --help -h                           Print this help message.
+    --mp=<num>                          The number of threads used in run.
+    --run_name=<run_name>               Select your run name.
+    --table_name=<table_name>           Target specific table.
+"""
+
+from subprocess import check_call
+from subprocess import Popen, PIPE
+import concurrent.futures
+import traceback
+import yaml
+from docopt import docopt
+import os
+from datetime import datetime
+import xml.etree.ElementTree as ET
+import time
+import pandas as pd
+import logging
+import glob
+from pathlib import Path
+from tqdm import tqdm
+import re
+import gdstk
+import errno
+
+from collections import defaultdict
+
+
+SUPPORTED_TC_EXT = "gds"
+SUPPORTED_SW_EXT = "yaml"
+RULE_LAY_NUM = 10000
+PATH_WIDTH = 0.01
+RULE_STR_SEP = "--"
+ANALYSIS_RULES = ["pass_patterns", "fail_patterns", "false_negative", "false_positive"]
+
+
+def get_unit_test_coverage(gds_file):
+    """
+    This function is used for getting all test cases available inside a single test table.
+    Parameters
+    ----------
+    gds_file : str
+        Path string to the location of unit test cases path.
+    Returns
+    -------
+    list
+        A list of unique rules found.
+    """
+    # Get rules from gds
+    rules = []
+
+    # layer num of rule text
+    lay_num = 11
+    # layer data type of rule text
+    lay_dt = 222
+
+    # Getting all rules names from testcase
+    library = gdstk.read_gds(gds_file)
```

Comment:
any reason we don't use `klayout` for that instead of gdstk?

Reply:
@proppy gdstk is extremely fast and actually has a very thin layer of code from GDS to memory. I prefer to avoid the use of klayout in simple tasks. The only reason I got to klayout is that it has more functionality.

Reply:
yes, the problem in the base database model for klayout. I have investigated the tool in and out. That's why I prefer all the work that is done by @joamatab from Google-X team. They are working gdsfactory which is much better than klayout in performance and speed. And actually, we are moving for example the PCells to use gdsfactory. I have a lot of details about this, but here is not the right place to discuss this.

https://github.com/gdsfactory/gdsfactory

Skywaters 130nm PDK based on GDS Factory:
https://github.com/efabless/sky130_klayout_pdk

GF180MCU PDK based on GDS Factory:
https://github.com/mabrains/globalfoundries-pdk-libs-gf180mcu_fd_pr/tree/Pcells_gdsfactry_regenerating

Reply:
my understanding is that gdsfactory is also exploring using klayout instead of gdstk as a backend, see: https://github.com/gdsfactory/kfactory

The work on gdsfactory with sky130/gf180mcu pdk is exciting but I agree it's unrelated to the topic discussed here.

In my experience using klayout for "rendering" is faster than gdstk, see:
https://colab.research.google.com/gist/proppy/7dac54b5aded7b05164791639d4bcae3/klayout-0-28-load_layout-1146.ipynb so I'd be curious to understand better the performance trade-off of using the bare python module gds parsing.

maybe @joamatab and @tvt173 could share their perspective here.

Reply:
@proppy What do you mean by "rendering"?

Klayout is a great editing tool. But not really fast when it come to script-able manipulation.

Reply:
@proppy Regardless of the discussion above, Could we please move on with merging the PR as I see this another topic for discussion?

Reply:
I have added #11 for further discussion.

Reply:
> @proppy What do you mean by "rendering"?

I mean parsing gds and rendering them as png, see the attached notebook:
https://colab.research.google.com/gist/proppy/7dac54b5aded7b05164791639d4bcae3/klayout-0-28-load_layout-1146.ipynb 

In comparison for large design gdstk takes a lot more time.

Reply:
> Could we please move on with merging the PR as I see this another topic for discussion?

Let's keep the discussion about "merging the PR" in the main comment, rather than nested here.